src/Orleans.CodeGenerator/SerializerGenerator.cs

```diff
@@ -479,52 +477,39 @@ private static void AddSerializationMembers(ISerializableTypeDescription type, L
 
             AddSerializationCallbacks(type, instanceParam, "OnDeserializing", body);
 
-            // C#: int id = 0;
-            body.Add(LocalDeclarationStatement(
-                VariableDeclaration(
-                    PredefinedType(Token(SyntaxKind.IntKeyword)),
-                    SingletonSeparatedList(VariableDeclarator(idVar.Identifier)
-                        .WithInitializer(EqualsValueClause(LiteralExpression(SyntaxKind.NumericLiteralExpression, Literal(0))))))));
-
-            // C#: Field header = default;
-            body.Add(LocalDeclarationStatement(
-                VariableDeclaration(
-                    libraryTypes.Field.ToTypeSyntax(),
-                    SingletonSeparatedList(VariableDeclarator(headerVar.Identifier)
-                        .WithInitializer(EqualsValueClause(LiteralExpression(SyntaxKind.DefaultLiteralExpression)))))));
-
-            int emptyBodyCount = 2;
-
-            if (type.SupportsPrimaryConstructorParameters)
+            int emptyBodyCount;
+            var normalMembers = members.FindAll(m => !m.IsPrimaryConstructorParameter);
+            if (members.Count == 0 || normalMembers.Count == 0 && !type.SupportsPrimaryConstructorParameters)
             {
-                body.Add(WhileStatement(LiteralExpression(SyntaxKind.TrueLiteralExpression), Block(GetDeserializerLoopBody(members.Where(m => m.IsPrimaryConstructorParameter)))));
-                body.Add(ExpressionStatement(AssignmentExpression(SyntaxKind.SimpleAssignmentExpression, IdentifierName(idVar.Identifier), LiteralExpression(SyntaxKind.NumericLiteralExpression, Literal(0)))));
-
-                body.Add(
-                    IfStatement(
-                        IdentifierName(headerVar.Identifier).Member("IsEndBaseFields"),
-                        WhileStatement(LiteralExpression(SyntaxKind.TrueLiteralExpression), Block(GetDeserializerLoopBody(members.Where(m => !m.IsPrimaryConstructorParameter))))));
+                // C#: reader.ConsumeEndBaseOrEndObject();
```

Comment:
The name is a little counterintuitive since it consumes up to and including the base or end object

Reply:
Yes, the xml docs on the method explain this also. But generally it's expected to be a no-op (just consume the EndBase/EndObject tag).